text/0000-builder-spec.md

```diff
@@ -0,0 +1,282 @@
+  samples/ruby-bundler          0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/ruby-bundler
+  samples/hello-universe        0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-universe
+  samples/hello-world           0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-world
+  samples/hello-moon            0.0.1          https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon
+
+Detection Order:
+ ├ Group #1:
+ │  └ samples/java-maven@0.0.1
+ ├ Group #2:
+ │  └ samples/kotlin-gradle@0.0.1
+ ├ Group #3:
+ │  └ samples/ruby-bundler@0.0.1
+ └ Group #4:
+    └ samples/hello-universe@0.0.1
+       └ Group #1:
+          ├ samples/hello-world@0.0.1
+          └ samples/hello-moon@0.0.1
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+## Structure of Builders
+### Files/Directories
+A builder is composed of at least the following directories/files:
+```
+- /cnb/buildpacks/...<buildpack ID>/<buildpack version>/
+- /cnb/lifecycle/<lifecycle binaries>
+- /cnb/order.toml
+- /cnb/stack.toml
+```
+
+### Env Vars/Labels
+A builder's environment is the build-time environment of the stack, and as such, requires all of the [build image specifications][build-image-specs].
+
+Additionally, a builder requires:
+* The image config's WorkingDir should be set to a directory in the builder
+* The image config's Env field has the environment variable `CNB_APP_DIR` set to the application directory of the build environment
+* The image config's Env field has the environment variable `CNB_LAYERS_DIR` set to a directory in the builder
+* The image config's Env field has the environment variable `CNB_PLATFORM_DIR` set to a directory in the builder
+* The image config's Env field has the environment variable `CNB_PLATFORM_DIR` set to a directory in the builder
+* The image config's Env field has the environment variable `SERVICE_BINDING_ROOT` set to a directory in the builder, if any services are bound
+* The image config's Labels field has the label io.buildpacks.builder.api, set to a string (defaulting to `0.1`)
+* The image config's Labels field has the label io.buildpacks.builder.metadata, set to a JSON object representing [Builder Metadata](#metadata)
+* The image config's Labels field has the label io.buildpacks.buildpack.order, set to a JSON object representing an [Order](#order)
+* The image config's Labels field has the label io.buildpacks.buildpack.layers, set to a JSON object representing the [buildpack layers](#layers)
+
+If the builder contains an optional [lifecycle descriptor file][lifecycle-descriptor-rfc], it also requires:
+* The image config's Labels field has the label io.buildpacks.lifecycle.version, set to the lifecycle version
+* The image config's Labels field has the label io.buildpacks.lifecycle.apis, set to a JSON object representing the [lifecycle apis](#lifecycle-apis)
```

Comment:
I'm trying to comprehend how a platform would be able to use a lifecycle without know it's API information given that this wording makes all that information optional. Was that discussed somewhere?

Reply:
Good catch. I added these requirements in, after @ekcasey reviewed it [here](https://github.com/buildpacks/rfcs/pull/116#discussion_r489611119). It sounds from the [lifecycle-descriptor rfc](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md#backwards-compatibility) that the the metadata for older versions was contained in the `io.buildpacks.builder.metadata` label:
> Currently lifecycle API and verison information is contained in the io.buildpacks.builder.metadata label. 

But, I removed that from the `io.buildpacks.builder.metadata` label after @ekcasey noted that we shouldn't have the old style in the new requirements. I'll update that to reflect that we assume those labels will be present.